posthog/api/process.py

```diff
@@ -0,0 +1,109 @@
+import structlog
+from typing import Any, Dict, List, Optional, cast
+
+from pydantic import BaseModel
+from rest_framework.exceptions import ValidationError
+
+from posthog.clickhouse.client import sync_execute
+from posthog.clickhouse.query_tagging import tag_queries
+from posthog.hogql.database.database import create_hogql_database, serialize_database
+from posthog.hogql.metadata import get_hogql_metadata
+from posthog.hogql.modifiers import create_default_modifiers_for_team
+from posthog.hogql_queries.query_runner import get_query_runner
+from posthog.models import Team
+from posthog.queries.time_to_see_data.serializers import SessionEventsQuerySerializer, SessionsQuerySerializer
+from posthog.queries.time_to_see_data.sessions import get_session_events, get_sessions
+from posthog.schema import HogQLMetadata
+from posthog.settings import CLICKHOUSE_CLUSTER
+from statshog.defaults.django import statsd
+
+logger = structlog.get_logger(__name__)
+
+QUERY_WITH_RUNNER = [
+    "LifecycleQuery",
+    "TrendsQuery",
+    "WebOverviewQuery",
+    "WebTopSourcesQuery",
+    "WebTopClicksQuery",
+    "WebTopPagesQuery",
+    "WebStatsTableQuery",
+]
+QUERY_WITH_RUNNER_NO_CACHE = ["EventsQuery", "PersonsQuery", "HogQLQuery", "SessionsTimelineQuery"]
+
+
+def _unwrap_pydantic(response: Any) -> Dict | List:
+    if isinstance(response, list):
+        return [_unwrap_pydantic(item) for item in response]
+
+    elif isinstance(response, BaseModel):
+        resp1: Dict[str, Any] = {}
+        for key in response.__fields__.keys():
+            resp1[key] = _unwrap_pydantic(getattr(response, key))
+        return resp1
+
+    elif isinstance(response, dict):
+        resp2: Dict[str, Any] = {}
+        for key in response.keys():
+            resp2[key] = _unwrap_pydantic(response.get(key))
+        return resp2
+
+    return response
+
+
+def _unwrap_pydantic_dict(response: Any) -> Dict:
+    return cast(dict, _unwrap_pydantic(response))
+
+
+def process_query(
```

Comment:
I think moving this out of query.py is good, though I kind of expect every `.py` file under `api` to expose some route. This only exposes utils. The convention so far has been to use various `utils.py` style files for shared code. Perhaps this could also be `query_utils.py` or so? I don't have a very strong case to make here, just that it felt weird that this doesn't actually expose any API routes.

Reply:
For me utils are more like reusable small things, whereas this is core to the endpoint. Maybe I'll move it to some package away from the route-exposing py files.